opm/autodiff/FlowMain.hpp

```diff
@@ -0,0 +1,673 @@
+#include <opm/autodiff/RedistributeDataHandles.hpp>
+#include <opm/autodiff/moduleVersion.hpp>
+
+#include <opm/core/utility/share_obj.hpp>
+
+#include <opm/parser/eclipse/OpmLog/OpmLog.hpp>
+#include <opm/parser/eclipse/OpmLog/StreamLog.hpp>
+#include <opm/parser/eclipse/OpmLog/CounterLog.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/Parser/Parser.hpp>
+#include <opm/parser/eclipse/Parser/ParseMode.hpp>
+#include <opm/parser/eclipse/EclipseState/checkDeck.hpp>
+#include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include <memory>
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+#include <numeric>
+#include <cstdlib>
+
+
+
+
+namespace Opm
+{
+
+    /// This class encapsulates the setup and running of
+    /// a simulator based on an input deck.
+    template <class Implementation, class Grid, class Simulator>
+    class FlowMainBase
+    {
+    public:
+
+
+        /// This is the main function of Flow.
+        /// It runs a complete simulation, with the given grid and
+        /// simulator classes, based on user command-line input.  The
+        /// content of this function used to be in the main() function of
+        /// flow.cpp.
+        int execute(int argc, char** argv)
+        try {
+            // Setup.
+            asImpl().setupParallelism(argc, argv);
+            asImpl().printStartupMessage();
+            const bool ok = asImpl().setupParameters(argc, argv);
+            if (!ok) {
+                return EXIT_FAILURE;
+            }
+            asImpl().setupOutput();
+            asImpl().readDeckInput();
+            asImpl().setupGridAndProps();
+            asImpl().setupState();
+            asImpl().distributeData();
+            asImpl().setupOutputWriter();
+            asImpl().setupLinearSolver();
+            asImpl().createSimulator();
+
+            // Run.
+            return asImpl().runSimulator();
+        }
+        catch (const std::exception &e) {
+            std::cerr << "Program threw an exception: " << e.what() << "\n";
+            return EXIT_FAILURE;
+        }
+
+
+
+    protected:
+
+
+
+
+
+        // ------------   Types   ------------
+
+
+        typedef BlackoilPropsAdFromDeck FluidProps;
+        typedef FluidProps::MaterialLawManager MaterialLawManager;
+        typedef typename Simulator::ReservoirState ReservoirState;
+
+
+        // ------------   Data members   ------------
+
+
+        // The comments indicate in which method the
+        // members first occur.
+
+        // setupParallelism()
+        bool output_cout_ = false;
+        bool must_distribute_ = false;
+        // setupParameters()
+        parameter::ParameterGroup param_;
+        // setupOutput()
+        bool output_to_files_ = false;
+        std::string output_dir_ = "output";
```

Comment:
I think that
```
std::string output_dir_("output");
```
should work as well, that may be allowed by your compiler? I'll change this and fire off a build test on Jenkins to see if it accepts it.